custom/openclinica/management/commands/odm_to_app.py

```diff
@@ -0,0 +1,198 @@
+"""
+Create a template app from ODM-formatted OpenClinica study metadata
+"""
+from lxml import etree
+from django.core.management.base import BaseCommand
+from corehq.apps.app_manager.const import APP_V2
+from corehq.apps.app_manager.models import Application, Module, OpenCaseAction
+from corehq.apps.app_manager.xform_builder import XFormBuilder
+from custom.openclinica.utils import odm_nsmap
+
+
+# Map ODM data types to XForm data types
+DATA_TYPES = {
+    'text': 'string',
+    'date': 'date',
+    'integer': 'int',
+
+}
+
+
+class StudyObject(object):
+    """
+    Base class of objects defined from ODM metadata
+    """
+    def __init__(self, defn, meta):
+        """
+        Initialize
+
+        :param defn: This object's definition
+        :param meta: Study metadata
+        """
+        self.defn = defn
+        self.meta = meta
+
+
+class Study(StudyObject):
+
+    def __init__(self, defn, meta):
+        super(Study, self).__init__(defn, meta)
+        self.oid = defn['OID']
+        self.name = defn.xpath('./odm:GlobalVariables/odm:StudyName', namespaces=odm_nsmap)[0].text
+        self._app = None
+
+    def iter_events(self):
+        for se_ref in self.meta.xpath('./odm:Protocol/odm:StudyEventRef', namespaces=odm_nsmap):
+            se_oid = se_ref['StudyEventOID']
+            se_def = self.meta.xpath('./odm:StudyEventDef[@OID="{}"]'.format(se_oid), namespaces=odm_nsmap)[0]
+            yield StudyEvent(se_def, self.meta)
+
+    @staticmethod
+    def new_subject_module():
+        module = Module.new_module('Study Subjects', None)
+        module.case_type = 'subject'
+        reg_form = module.new_form('Register Subject')
+        reg_form.actions.open_case = OpenCaseAction(name_path="/data/name", external_id=None)
+        reg_form.actions.open_case.condition.type = 'always'
+        return module
+
+    def get_app(self, domain_name, app_name, version=APP_V2):
+        if self._app is None:
+            self._app = Application.new_app(domain_name, app_name, application_version=version)
+            self._app.name = self.name
+            self._app.add_module(self.new_subject_module())
+            for event in self.iter_events():
+                module = event.get_module()
+                for study_form in event.iter_forms():
+                    study_form.add_form_to_module(module)
+                self._app.add_module(module)
+        return self._app
```

Comment:
This `self._app` member appears to be unused. It's not referenced outside of `get_app()`, and `get_app()` is only called once in the `Command` at the bottom of the file.